packages/component-base/src/focus-utils.js

```diff
@@ -0,0 +1,212 @@
+  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort
+  const len = elements.length;
+  if (len < 2) {
+    return elements;
+  }
+  const pivot = Math.ceil(len / 2);
+  const left = sortElementsByTabIndex(elements.slice(0, pivot));
+  const right = sortElementsByTabIndex(elements.slice(pivot));
+
+  return mergeSortByTabIndex(left, right);
+}
+
+/**
+ * Searches for nodes that are tabbable and adds them to the `result` array.
+ * Returns if the `result` array needs to be sorted by tabindex.
+ *
+ * @param {Node} node The starting point for the search; added to `result` if tabbable.
+ * @param {HTMLElement[]} result
+ * @return {boolean}
+ * @private
+ */
+function collectFocusableNodes(node, result) {
+  // If the node is not an element or hidden, no need to traverse children.
+  if (node.nodeType !== Node.ELEMENT_NODE || isElementHidden(node)) {
+    return false;
+  }
+  const element = /** @type {HTMLElement} */ (node);
+  const tabIndex = normalizeTabIndex(element);
+  let needsSort = tabIndex > 0;
+  if (tabIndex >= 0) {
+    result.push(element);
+  }
+
+  // In ShadowDOM v1, tab order is affected by the order of distribution.
+  // E.g. getTabbableNodes(#root) in ShadowDOM v1 should return [#A, #B];
+  // in ShadowDOM v0 tab order is not affected by the distribution order,
+  // in fact getTabbableNodes(#root) returns [#B, #A].
+  //  <div id="root">
+  //   <!-- shadow -->
+  //     <slot name="a">
+  //     <slot name="b">
+  //   <!-- /shadow -->
+  //   <input id="A" slot="a">
+  //   <input id="B" slot="b" tabindex="1">
+  //  </div>
+  let children;
+  if (element.localName === 'slot') {
+    children = element.assignedNodes({ flatten: true });
+  } else {
+    // Use shadow root if possible, will check for distributed nodes.
+    children = (element.shadowRoot || element).children;
+  }
+  if (children) {
+    for (let i = 0; i < children.length; i++) {
+      // Ensure method is always invoked to collect tabbable children.
+      needsSort = collectFocusableNodes(children[i], result) || needsSort;
+    }
+  }
+  return needsSort;
+}
+
+/**
+ * Returns true if the element is focusable, otherwise false.
+ *
+ * The list of focusable elements is taken from http://stackoverflow.com/a/1600194/4228703.
+ * However, there isn't a definite list, it's up to the browser.
+ * The only standard we have is DOM Level 2 HTML https://www.w3.org/TR/DOM-Level-2-HTML/html.html,
+ * according to which the only elements that have a `focus()` method are:
+ * - HTMLInputElement
+ * - HTMLSelectElement
+ * - HTMLTextAreaElement
+ * - HTMLAnchorElement
+ *
+ * This notably omits HTMLButtonElement and HTMLAreaElement.
+ * Referring to these tests with tabbables in different browsers
+ * http://allyjs.io/data-tables/focusable.html
+ *
+ * @param {HTMLElement} element
+ * @return {boolean}
+ */
+export function isElementFocusable(element) {
```

Comment:
**note:** Renamed `isFocusable` to `isElementFocusable`.

https://github.com/vaadin/web-components/blob/699547e5b2976e1ad0838414bb478c752d31245b/packages/vaadin-overlay/src/vaadin-focusables-helper.js#L37-L54